File: apps/lifenavigator-core-api/app/services/response_review.py
"""Advisor-response report review workflow — audit finding R-3 / B-23.

CAPABILITY BOUNDARY
-------------------
Every method here requires `PlatformAccess.can_review_advisor_responses`. That capability is
deliberately narrower than `is_admin`: it authorizes *these* operations and nothing else. Every
other admin endpoint keeps calling `is_admin`, so review access cannot widen into administration.

WHAT A REVIEWER CANNOT DO — enforced by absence, not by a check
---------------------------------------------------------------
There is no delete method. There is no method that writes to the report's evidence columns. There
is no method touching models, prompts, policies, feature flags, cohorts or invitations. A
capability you cannot express is stronger than one you check for.

DATA MINIMIZATION
-----------------
`list_reports` returns queue columns ONLY — no question or response snapshot. A reviewer scanning a
queue has no need to read every reporter's question, and returning them would turn the queue into a
bulk disclosure surface. Snapshots require opening one report deliberately.

`llm_response_raw` is not in the report table at all (ADR-002 / slice 1), so it cannot leak here.
"""
from __future__ import annotations

import uuid
from typing import Any, Optional

# Queue columns. Deliberately excludes question_snapshot / response_snapshot / explanation.
LIST_COLUMNS = (
    "report_id,turn_id,category,severity,review_status,assigned_to,created_at,"
    "deployment_version,prompt_version,model_name,duplicate_of"
)

# Detail columns. Adds the preserved evidence a reviewer needs to investigate ONE report.
DETAIL_COLUMNS = LIST_COLUMNS + ",explanation,question_snapshot,response_snapshot,citation_refs,retrieval_channels,model_provider,resolution_notes,conversation_id"

VALID_SEVERITIES = ("unclassified", "low", "medium", "high", "critical")

# Explicit transition table. An arbitrary status string, or a transition not listed here, is
# refused — a review workflow that accepts any string is not a workflow.
ALLOWED_TRANSITIONS: dict[str, tuple[str, ...]] = {
    "new": ("investigating", "dismissed", "duplicate"),
    "investigating": ("resolved", "dismissed", "duplicate"),
    # Reopening is permitted but ONLY back to investigating, and the caller must supply a reason.
    "resolved": ("investigating",),
    "dismissed": ("investigating",),
    "duplicate": ("investigating",),
}
REOPEN_FROM = ("resolved", "dismissed", "duplicate")

MAX_PAGE_SIZE = 100


class ReviewForbidden(Exception):
    """Caller lacks the advisor_response_reviewer capability."""


class ReviewRejected(Exception):
    """The requested change is not permitted by the workflow contract."""


class ResponseReviewService:
    def __init__(self, supabase: Any, access: Any) -> None:
        self._sb = supabase
        self._access = access

    # ── authorization ───────────────────────────────────────────────────────
    async def _authorize(self, ctx: Any, email: Optional[str], endpoint: str) -> str:
        """Authorize and audit. Denials are audited too — a denied attempt is the earliest signal
        that someone is probing an endpoint they should not know about."""
        if not self._access.can_review_advisor_responses(email):
            await self._access.log_admin_access(ctx, email, endpoint, "denied")
            raise ReviewForbidden("reviewer capability required")
        await self._access.log_admin_access(ctx, email, endpoint, "granted")
        return (email or "").lower()

    # ── read ────────────────────────────────────────────────────────────────
    async def list_reports(
        self, ctx: Any, email: Optional[str], *, filters: Optional[dict[str, str]] = None,
        limit: int = 25, offset: int = 0,
    ) -> dict[str, Any]:
        await self._authorize(ctx, email, "/v1/admin/response-reports")
        bounded = max(1, min(int(limit or 25), MAX_PAGE_SIZE))
        allowed = {"category", "review_status", "severity", "deployment_version",
                   "prompt_version", "model_name", "assigned_to"}
        query = {k: f"eq.{v}" for k, v in (filters or {}).items() if k in allowed and v}
        rows = await self._sb.select(
            "advisor_response_reports", columns=LIST_COLUMNS, filters=query,
            limit=bounded, order="created_at.desc", schema="feedback",
        )
        return {"reports": rows or [], "limit": bounded, "offset": max(0, int(offset or 0))}

    async def get_report(self, ctx: Any, email: Optional[str], report_id: str) -> dict[str, Any]:
        await self._authorize(ctx, email, f"/v1/admin/response-reports/{report_id}")
        rows = await self._sb.select(
            "advisor_response_reports", columns=DETAIL_COLUMNS,
            filters={"report_id": f"eq.{report_id}"}, schema="feedback",
        )
        if not rows:
            raise ReviewRejected("not found")
        events = await self._sb.select(
            "advisor_report_events",
            columns="event_id,action,actor_email,previous_value,new_value,note,created_at",
            filters={"report_id": f"eq.{report_id}"}, order="created_at.desc", schema="feedback",
        )
        return {"report": rows[0], "audit_history": events or []}

    # ── write — the ONLY mutable surface ────────────────────────────────────
    async def update_review(
        self, ctx: Any, email: Optional[str], report_id: str, *,
        assigned_to: Optional[str] = None, severity: Optional[str] = None,
        review_status: Optional[str] = None, note: Optional[str] = None,
        duplicate_of: Optional[str] = None, escalate: bool = False,
    ) -> dict[str, Any]:
        actor = await self._authorize(ctx, email, f"/v1/admin/response-reports/{report_id}")

        current = await self._sb.select(
            "advisor_response_reports", columns="report_id,review_status,severity,assigned_to",
            filters={"report_id": f"eq.{report_id}"}, schema="feedback",
        )
        if not current:
            raise ReviewRejected("not found")
        row = current[0]

        # Only these four columns are ever written. Evidence columns are absent from `patch` by
        # construction, so no request shape can reach them.
        patch: dict[str, Any] = {}
        events: list[dict[str, Any]] = []

        if assigned_to is not None:
            patch["assigned_to"] = assigned_to
            events.append(self._event(report_id, actor, "assigned", row.get("assigned_to"), assigned_to, note))

        if severity is not None:
            if severity not in VALID_SEVERITIES:
                raise ReviewRejected("invalid severity")
            patch["severity"] = severity
            events.append(self._event(report_id, actor, "severity_changed", row.get("severity"), severity, note))

        if review_status is not None:
            prior = str(row.get("review_status") or "new")
            allowed = ALLOWED_TRANSITIONS.get(prior, ())
            if review_status not in allowed:
                raise ReviewRejected(f"transition {prior} -> {review_status} is not permitted")
            if prior in REOPEN_FROM and not (note or "").strip():
                raise ReviewRejected("reopening requires a reason")
            patch["review_status"] = review_status
            if review_status == "duplicate":
                if not duplicate_of:
                    raise ReviewRejected("duplicate requires duplicate_of")
                patch["duplicate_of"] = duplicate_of
                events.append(self._event(report_id, actor, "duplicate_linked", None, duplicate_of, note))
            events.append(self._event(report_id, actor, "status_changed", prior, review_status, note))

        if note and not events:
            events.append(self._event(report_id, actor, "note_added", None, None, note))

        if escalate:
            # The reviewer records the escalation; the reviewer capability itself performs NO
            # containment. Disabling a model, changing a prompt or touching a flag is reserved for
            # the beta incident owner / platform administrator.
            events.append(self._event(report_id, actor, "escalated", None, None,
                                      note or "escalated to beta incident owner"))

        if not patch and not events:
            raise ReviewRejected("no change requested")

        if patch:
            await self._sb.update(
                "advisor_response_reports", patch,
                filters={"report_id": f"eq.{report_id}"}, schema="feedback",
            )
        for ev in events:
            await self._sb.insert("advisor_report_events", ev, schema="feedback")
        return {"report_id": report_id, "updated": sorted(patch), "events": len(events)}

    @staticmethod
    def _event(report_id: str, actor: str, action: str, prev: Any, new: Any,
               note: Optional[str]) -> dict[str, Any]:
        return {
            "event_id": str(uuid.uuid4()), "report_id": report_id, "actor_email": actor,
            "action": action,
            "previous_value": None if prev is None else str(prev),
            "new_value": None if new is None else str(new),
            "note": note or None,
        }

File: apps/web/src/lib/admin/responseReview.ts
/**
 * Typed reviewer client for the advisor-response review queue (R-3 / B-27).
 *
 * THE SERVER REMAINS AUTHORITATIVE. This client cannot grant capability, cannot bypass a transition
 * rule, and deliberately exposes NO method for deleting a report, editing evidence, changing a
 * reporter/tenant/turn/category, touching feature flags, prompts, models, cohorts, or suspending a
 * user. Those operations are absent rather than guarded — an API you cannot call is stronger than
 * one you check for.
 *
 * A 403 is surfaced as `forbidden`, never as an empty queue. Rendering "no reports" for a denied
 * reviewer would hide an authorization failure behind a plausible empty state.
 */

export const REVIEW_STATUSES = [
  'new',
  'investigating',
  'resolved',
  'dismissed',
  'duplicate',
] as const;
export type ReviewStatus = (typeof REVIEW_STATUSES)[number];

export const SEVERITIES = ['unclassified', 'low', 'medium', 'high', 'critical'] as const;
export type Severity = (typeof SEVERITIES)[number];

/** Mirrors the server transition table. Advisory only — the server re-validates every transition. */
export const ALLOWED_TRANSITIONS: Record<ReviewStatus, ReviewStatus[]> = {
  new: ['investigating', 'dismissed', 'duplicate'],
  investigating: ['resolved', 'dismissed', 'duplicate'],
  resolved: ['investigating'],
  dismissed: ['investigating'],
  duplicate: ['investigating'],
};

/** Reopening a terminal state requires a reason; the server refuses without one. */
export const REOPEN_FROM: ReviewStatus[] = ['resolved', 'dismissed', 'duplicate'];

export function nextStatuses(current: ReviewStatus): ReviewStatus[] {
  return ALLOWED_TRANSITIONS[current] ?? [];
}

export function requiresReason(current: ReviewStatus): boolean {
  return REOPEN_FROM.includes(current);
}

/** Queue row — minimized by the API. No question/response snapshot, no explanation. */
export interface ReportSummary {
  report_id: string;
  turn_id: string;
  category: string;
  severity: Severity;
  review_status: ReviewStatus;
  assigned_to: string | null;
  created_at: string;
  deployment_version: string | null;
  prompt_version: string | null;
  model_name: string | null;
  duplicate_of: string | null;
}

export interface AuditEvent {
  event_id: string;
  action: string;
  actor_email: string;
  previous_value: string | null;
  new_value: string | null;
  note: string | null;
  created_at: string;
}

/** Detail — preserved evidence, loaded only when a reviewer deliberately opens one report. */
export interface ReportDetail extends ReportSummary {
  explanation: string | null;
  question_snapshot: string | null;
  response_snapshot: string | null;
  citation_refs: unknown[] | null;
  retrieval_channels: unknown[] | null;
  model_provider: string | null;
  resolution_notes: string | null;
  conversation_id: string | null;
}

export interface QueueFilters {
  category?: string;
  review_status?: string;
  severity?: string;
  deployment_version?: string;
  prompt_version?: string;
  model_name?: string;
  assigned_to?: string;
}

export type ReviewOutcome<T> =
  | { kind: 'ok'; data: T }
  | { kind: 'forbidden' }
  | { kind: 'not_found' }
  | { kind: 'rejected'; message: string }
  | { kind: 'error'; status: number }
  | { kind: 'network_error' };

export interface ReviewUpdate {
  assigned_to?: string;
  severity?: Severity;
  review_status?: ReviewStatus;
  note?: string;
  duplicate_of?: string;
  escalate?: boolean;
}

export const MAX_PAGE_SIZE = 100;

interface Options {
  fetchImpl?: typeof fetch;
}

export async function listReports(
  filters: QueueFilters = {},
  page: { limit?: number; offset?: number } = {},
  { fetchImpl = fetch }: Options = {}
): Promise<ReviewOutcome<{ reports: ReportSummary[]; limit: number; offset: number }>> {
  const qs = new URLSearchParams();
  for (const [k, v] of Object.entries(filters)) if (v) qs.set(k, v);
  qs.set('limit', String(Math.min(page.limit ?? 25, MAX_PAGE_SIZE)));
  if (page.offset) qs.set('offset', String(page.offset));
  return request(`/api/admin/response-reports?${qs}`, { fetchImpl });
}

export async function getReport(
  reportId: string,
  { fetchImpl = fetch }: Options = {}
): Promise<ReviewOutcome<{ report: ReportDetail; audit_history: AuditEvent[] }>> {
  return request(`/api/admin/response-reports/${encodeURIComponent(reportId)}`, { fetchImpl });
}

export async function updateReport(
  reportId: string,
  update: ReviewUpdate,
  { fetchImpl = fetch }: Options = {}
): Promise<ReviewOutcome<{ report_id: string; updated: string[]; events: number }>> {
  // Built field-by-field — never spread — so no caller object can smuggle an evidence column.
  const body: ReviewUpdate = {};
  if (update.assigned_to) body.assigned_to = update.assigned_to;
  if (update.severity) body.severity = update.severity;
  if (update.review_status) body.review_status = update.review_status;
  if (update.note) body.note = update.note;
  if (update.duplicate_of) body.duplicate_of = update.duplicate_of;
  if (update.escalate) body.escalate = true;

  return request(`/api/admin/response-reports/${encodeURIComponent(reportId)}`, {
    fetchImpl,
    init: {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    },
  });
}

async function request<T>(
  url: string,
  { fetchImpl, init }: { fetchImpl: typeof fetch; init?: RequestInit }
): Promise<ReviewOutcome<T>> {
  let res: Response;
  try {
    res = await fetchImpl(url, init);
  } catch {
    return { kind: 'network_error' };
  }
  // 403 is surfaced explicitly. Collapsing it into an empty result would hide an authorization
  // failure behind a plausible "no reports" screen.
  if (res.status === 401 || res.status === 403) return { kind: 'forbidden' };
  if (res.status === 404) return { kind: 'not_found' };
  if (res.status === 400) {
    const body = await safeJson(res);
    return { kind: 'rejected', message: readDetail(body) ?? 'That change was rejected.' };
  }
  if (res.status >= 200 && res.status < 300) {
    return { kind: 'ok', data: ((await safeJson(res)) ?? {}) as T };
  }
  return { kind: 'error', status: res.status };
}

async function safeJson(res: Response): Promise<Record<string, unknown> | null> {
  try {
    return (await res.json()) as Record<string, unknown>;
  } catch {
    return null;
  }
}

function readDetail(body: Record<string, unknown> | null): string | null {
  const d = body?.detail;
  return typeof d === 'string' && d ? d : null;
}
